src/lauxlib.js

```diff
@@ -644,19 +644,144 @@ const luaL_unref = function(L, t, ref) {
+** a first-line comment).
+*/
+const skipcomment = function(lf) {
+    let c = skipBOM(lf);
+    if (c === '#'.charCodeAt(0)) {  /* first line is a comment (Unix exec. file)? */
+        do {  /* skip first line */
+            c = getc(lf);
+        } while (c && c !== '\n'.charCodeAt(0));
+
+        return {
+            skipped: true,
+            c: getc(lf)  /* skip end-of-line, if present */
+        };
+    } else {
+        return {
+            skipped: false,
+            c: c
+        };
+    }
+};
+
+class LoadF {
+    constructor() {
+        this.n = NaN;  /* number of pre-read characters */
+        this.f = null;  /* file being read */
+        this.buff = WEB ? new Array(1024) : new Buffer(1024);  /* area for reading file */
+        this.pos = 0;  /* current position in file */
+        this.err = void 0;
     }
+}
+
+if (WEB) {
+    const getF = function(L, ud) {
+        let lf = ud;
+
+        if (lf.f !== null && lf.n > 0) {  /* are there pre-read characters to be read? */
+            let bytes = lf.n; /* return them (chars already in buffer) */
+            lf.n = 0;  /* no more pre-read characters */
+            lf.f = lf.f.slice(lf.pos);  /* we won't use lf.buff anymore */
+            return lf.buff.slice(0, bytes);
+        }
+
+        let f = lf.f;
+        lf.f = null;
+        return f;
+    };
+
+    getc = function(lf) {
+        return lf.pos < lf.f.length ? lf.f[lf.pos++] : null;
+    };
+
+    const luaL_loadfilex = function(L, filename, mode) {
+        let lf = new LoadF();
+        let fnameindex = lua.lua_gettop(L) + 1;  /* index of filename on the stack */
+        if (filename === null) {
+            throw new Error(`Can't read stdin in the browser`);
+        } else {
+            let jsfilename = lua.to_jsstring(filename);
+            lua.lua_pushliteral(L, `@${jsfilename}`);
+
+            let xhr = new XMLHttpRequest();
+            xhr.open("GET", jsfilename, false);
+            //xhr.responseType = "arraybuffer";
+            xhr.send();
+
+            if (xhr.status === 200) {
+                /* TODO: Synchronous xhr alway return a js string */
+                lf.f = lua.to_luastring(xhr.response);
+            } else {
+                lf.err = xhr.status;
+                return errfile(L, "open", fnameindex, { message: `${xhr.status}: ${xhr.statusText}` });
+            }
+        }
+        let com = skipcomment(lf);
+        /* check for signature first, as we don't want to add line number corrections in binary case */
+        if (com.c === lua.LUA_SIGNATURE.charCodeAt(0) && filename) {  /* binary file? */
+            /* no need to re-open in node.js */
+        } else if (com.skipped) { /* read initial portion */
+            lf.buff[lf.n++] = '\n'.charCodeAt(0);  /* add line to correct line numbers */
+        }
+        if (com.c !== null)
+            lf.buff[lf.n++] = com.c; /* 'c' is the first character of the stream */
```

Comment:
What sets/fills in lf.buff? I only see us setting `lf.f` and leaving lf.buff as an empty array.